### PR TITLE
[RFC] a new version of gb-vendor

### LINKED ISSUE #130: .gitignore files from vendor directory causing issues
Looks like some of my deps did not land in my upstream commit. Taking a closer look it seems the .gitignore files from the vendor directories are causing issues.  Checkout the following session:

Vendor the consul api:

```
confd $ gb vendor github.com/hashicorp/consul/api
```

Notice nothing new shows up in the vendor dir:

```
confd $ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   .gitignore

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	bin/

no changes added to commit (use "git add" and/or "git commit -a")
```

Find all .gitignore files

```
confd $ find ./ -name .gitignore
.//.gitignore
.//vendor/src/github.com/hashicorp/consul/.gitignore
.//vendor/src/github.com/hashicorp/consul/website/source/.gitignore
```

Remove them:

```
confd $ rm vendor/src/github.com/hashicorp/consul/.gitignore
confd $ rm vendor/src/github.com/hashicorp/consul/website/source/.gitignore
```

Check git status again. Notice that the vendor dir changes are picked up:

```
confd $ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)

	modified:   .gitignore
	modified:   vendor/src/github.com/hashicorp/consul (modified content)

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	bin/

no changes added to commit (use "git add" and/or "git commit -a")
```

This is really a problem with my git setup, but it might bite people down the road.

### PR DESCRIPTION
**Pull request by davecheney**
Updates #17 
Updates #29 
Updates #71 
Updates constabulary/gb-vendor#1
Updates constabulary/gb-vendor#2
Updates constabulary/gb-vendor#4
Updates constabulary/gb-vendor#5
Updates constabulary/gb-vendor#6
Updates constabulary/gb-vendor#7

Closes #130

This PR contains a new version of the `gb-vendor` plugin. The aim is to replace, in short order, the existing constabulary/gb-vendor plugin, which will then be removed.

This new `gb-vendor` does not rely on `go get`. 

## Usage

`gb-vendor` accepts subcommands. At the moment there are thee subcommand is `fetch`, `update` and `delete` (thanks to @davidkaya).
```
% gb vendor
Usage:
  gb delete [flags] [package] - deletes a local dependency
  gb fetch [flags] [package] - fetch a remote dependency
  gb update [flags] [package] - updates a local dependency
```
 `fetch` takes a single argument, an import path. Here is an example of using `gb vendor fetch`
```
% gb vendor fetch github.com/kr/fs
Cloning into '/tmp/gb-vendor-161722046'...
remote: Counting objects: 18, done.
remote: Total 18 (delta 0), reused 0 (delta 0), pack-reused 18
Unpacking objects: 100% (18/18), done.
Checking connectivity... done.
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/LICENSE, src: /tmp/gb-vendor-161722046/LICENSE)
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/Readme, src: /tmp/gb-vendor-161722046/Readme)
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/example_test.go, src: /tmp/gb-vendor-161722046/example_test.go)
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/filesystem.go, src: /tmp/gb-vendor-161722046/filesystem.go)
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/walk.go, src: /tmp/gb-vendor-161722046/walk.go)
copyfile(dst: /home/dfc/devel/demo4/vendor/src/github.com/kr/fs/walk_test.go, src: /tmp/gb-vendor-161722046/walk_test.go)
```
As usual the output is verbose, it will be toned down in later revisions.

Support for fetching revisions and branches will be added later.

~~Other commands like `delete` and `update` are planned, but not written at this time.~~ Thanks to @davidkaya for adding these.

- `delete` removes the source from `$PROJECT/vendor/src` and the reference from the manifest file.
- `update` replaces the source with the latest available from the head of the master branch.

## Manifest file
`gb-vendor` records dependencies fetched (and in the future, updated) in a manifest file, which is currently called ~~`$PROJECT/vendor/vendorfile`~~ `$PROJECT/vendor/manifest`. This is a json file because `encoding/json` comes for free with the standard library, and is somewhat human readable. 

Here is an example of the manifest file after the previous command
```
% jsonpp vendor/manifest
{
  "version": 0,
  "dependencies": [
    {
      "importpath": "github.com/kr/fs",
      "repository": "https://github.com/kr/fs/",
      "revision": "2788f0dbd16903de03cb8186e5c7d97b69ad387b",
      "branch": "master",
      "path": ""
    }
  ]
}
```
The contents of the manifest file are described in its godoc.

The manifest file is not part of `gb`, it is used only by `gb-vendor`.

## DVCS support

Only git and https:// is supported at the moment, additional DVCSs and schemes will be added as required.

## For discussion

- The name of the manifest file, ~~`$PROJECT/vendor/vendorfile`~~ `$PROJECT/vendor/manifest` will be fixed when this commit lands. This is your opportunity to bikeshed.

### CODE REVIEW

**Comment by stuart-warren** [comment]
Cool. I'd like to propose that we make use of the go-import html meta tags (https://golang.org/cmd/go/#hdr-Remote_import_paths) then fall back to suffixes (.git/.hg/etc) then to hardcoded prefixes like the present github one

meta tags are used at most of the repo sites including github and gopkg.in and help find the root of the repo if you try to do something silly like ``gb vendor fetch github.com/bitly/nsq/internal/app``

see how they parse the html headers with ``go get``
https://github.com/golang/go/blob/439b32936367c3efd0dadab48dd51202e1a510f1/src/cmd/go/discovery.go#L38

**Comment by kharin** [comment]
My 2 cents:

* I don't like the idea of making a new, "everything for everyone" wrapper for any possible DVCS out there. You might end up re-implementing everything "go get" does (people would expect that), while trying to implement what submodules/subtrees already do. I mean, good luck keeping up with feature requests. :)
* It's also kind of sad and ironic, that a build system built to lay an end to "lockfiles" ends up with "vendorfile" of its own. Git submodules _are_ lockfiles to me.

More importantly, I don't think vendoring is something that needs to be simplified at all.

**Comment by davecheney** [comment]
@stuart-warren i'll add support for vanity imports sometime (free free to add a bug), but I think it's more important to add support for things like `gb vendor fetch git+ssh://vpn.mycompany/...`, which will happen first.

Unless I've misunderstood what your asking, in which case, log a bug

**Comment by davidkaya** [comment]
@kharin 
Ad 1. Well, I think Dave wanted the `gb-vendor` to be better than the `go get` since the beginning even though it could mean "re-implementing" it. As far as I know there is, e.g., no planned support for versioning packages in `go get` in any of the upcoming releases, which is a huge drawback of the `go get`. And as for the keeping up with the feature requests, I think open source community exists for a reason and Dave is not alone here, so I don't think that will be a problem. :)

Ad 2. The vendorfile is a huge advantage. There is nothing better than having control over dependencies. Git submodules can't be used in most of the (company) build environments, since those generally don't have access to the Internet and that forces you to include the dependencies in the company's vcs (which includes excluding .git-style folders). And all of this could easily lead to loss of control over versions of the dependencies. The vendorfile gives you control over all of the depenendencies and let's you easily modify them.

**Comment by rnapier** [comment]
Echoing @stuart-warren, is there a reason to to support the `go get` lookup logic, at least for git? The current hardcoded approach seems unscalable (I'm using a gopkg.in repo, so already hit the issue), and that part of `go get` I think works well.

I'm happy to make this comment in the form of a PR (currently working on it, but wanted to touch base first).

**Comment by davecheney** [comment]
@kharin `gb-vendor` is completely optional, if submodules work for you then you can ignore that gb-vendor exists. 

`gb` doesn't care about git submodules, or even git, it only cares about files on disk laid out in a specific way. How that happens is outside it's area of concern, you can use svn externals, git sub_mumble_, or gb-vendor, a makefile, or something else.

**Comment by davecheney** [comment]
@rnapier ues, I'll support all the forms that `go get` supports, although I doubt people will use them. What people really want is `gb vendor fetch git://server.firewall.company.com/flobble/wilbble.git --branch twinkle-star`

**Comment by PSG-Luna** [comment]
`gb vendor fetch .` is also very very useful to just be able to grab all
the deps of the current project

**Comment by rnapier** [comment]
@davecheney I see the desire for full URLs (I don't have anything like that), but don't you think many people want non-github domains (without needing access to special URLs)? I wanted gopkg.in, since that's how https://github.com/go-fsnotify/fsnotify asks to be imported. We also have stuff from bitbucket.org.

**Comment by davecheney** [comment]
@rnapier, yup all those usual domains will come quick smart once this PR is landed. 

Adding support for vanity imports will be interesting as gb-vendor will deobfuscate those import paths, consider this example

```
  "dependencies": [
    {
      "importpath": "gopkg.in/fsnotify.v1",
      "repository": "https://github.com/go-fsnotify/fsnotify",
      "revision": "feedface",
      "branch": "master",
      "path": ""
    }
  ]
```

**Comment by davecheney** [comment]
@psg-luna

> `gb vendor fetch .` is also very very useful to just be able to grab all the deps of the current project

I want to go a little slow on globbing operators like this because dependency inclusion should be a thought provoking exercise, if you're a project like hugo or juju I want the project owner to have cause to consider "do I really need every single one of these dependencies"

**Comment by rnapier** [comment]
@davecheney Interesting about the demangling of redirection. I do like that "repository" and "importpath" are completely independent. OK, I will be patient and try moving forward by hand-loading the things I need for testing with gb.

Also, agree about your comment about "maybe imports should be thought about."

**Comment by PSG-Luna** [comment]
I agree. However, when starting a new workspace and adding projects to it,
vendoring the dependencies one by one is somewhat of a bother. It's a
onetime process however, so I don't mind it overly much.

**Comment by davecheney** [comment]
@rnapier @PSG-Luna I agree that starting or converting a project from scratch is a pain, and this runs counter to ideas of gb user adoption. I will look at those, but not til I get support for the usual repo locations added, and support for other forms of git urls, 'cos I know there are a bunch of people who want to use git+ssh:// etc.

**Comment by davecheney** [comment]
Failing build is due to a flaky test, #128. We'll fix that before merging.

**Comment by davecheney** [comment]
*Update*: I have renamed the name of the manifest file from `vendorfile` to `manifest`. This is mainly to reduce the stuttering of `vendor/vendorfile`

/cc @kelseyhightower

File: cmd/gb-vendor/vendor/manifest.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+)
+
+// gb-vendor manifest support
+
+// Manfest describes the layout of $PROJECT/vendor/vendorfile.
```
Comment (mmlb) [inline cmd/gb-vendor/vendor/manifest.go]: s/Manfest/Manifest/
s/vendorfile/manfiest/

File: cmd/gb-vendor/main.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	if err := command.Run(ctx, args); err != nil {
+		gb.Fatalf("command %q failed: %v", name, err)
+	}
+}
+
+const manifestfile = "manifest"
+
+// manifestFile returns $PROJECT/vendor/$vendorfile
```
Comment (mmlb) [inline cmd/gb-vendor/main.go]: s/$vendorfile/$manifestfile/

File: cmd/gb-vendor/vendor/manifest.go (hunk 1 above)
Comment (davecheney) [inline cmd/gb-vendor/vendor/manifest.go]: fixed

File: cmd/gb-vendor/main.go (hunk 2 above)
Comment (davecheney) [inline cmd/gb-vendor/main.go]: fixed

**Comment by andrewchambers** [comment]
Thanks Dave et al.

**Pull request merged**